### DISCUSSION TITLE
jvm/how-run-java-code

### DISCUSSION THREAD

**Discussion by giscus** [itwanger/tobebetterjavaer-giscus#153] [Announcements]
# jvm/how-run-java-code

Java程序员进阶之路，小白的零基础Java教程，从入门到进阶，JVM到底是如何运行Java代码的？

https://tobebetterjavaer.com/jvm/how-run-java-code.html

**Comment by chaofengdev** [comment]
为什么没人说话？写的真好。

**Comment by chaofengdev** [comment]
我有一点不太确定：
“当 calculate 方法执行完成后，对应的栈帧将从虚拟机栈中弹出，方法执行的结果会被压入 main 栈帧中的操作数栈中，而方法返回地址被重置到 main 线程的 PC 寄存器中，以便于后续字节码执行引擎从 PC 寄存器中获取下一条命令的地址。”
在这里，我可以将calculate方法的栈帧里的`方法返回地址`理解为它的调用者`main`方法`在执行完被调方法后的字节码指令的地址`吗？

**Comment by chaofengdev** [comment]
去查阅了一点资料，在这里做了点笔记：
`JVM-栈帧中的方法返回地址(return address)`
 存放的是调用该方法的pc寄存器的值。

 一个方法的结束，有两种方式：
1.正常执行完成
2.出现未处理的异常，非正常退出
`无论通过哪种方式退出，在方法退出后都返回到该方法被调用的位置。方法正常退出时，调用者的pc计数器的值作为返回地址，即调用该方法的指令的下一条指令的地址。而通过异常退出的，返回地址是要通过异常表来确定，栈帧中一般不会保存这部分信息。`
本质上，方法的退出就是当前栈帧出栈的过程。此时，需要恢复上层方法的局部变量表、操作数栈、将返回值压入调用者栈帧的操作数栈、设置PC寄存器值等，让调用者方法继续执行下去。
 
 正常完成出口和异常完成出口的区别在于：通过异常完成出口退出的不会给他的上层调用者产生任何的返回值。

**Comment by WuHututu** [comment]
最后一句详细一点应该是：
执行引擎中的解释器根据程序计数器中记录的当前字节码指令的偏移量，去元空间（方法区）中找到当前正在执行的
方法的完整字节码指令序列，在获取到指令之后，通过解释器解释为对应的机器指令，最终由 CPU 进行执行。
